File: trunk/plugins/com.googlecode.contraildb.core/src/com/googlecode/contraildb/core/storage/ObjectStorage.java
package com.googlecode.contraildb.core.storage;

import java.io.ByteArrayInputStream;
import java.io.DataInputStream;
import java.io.IOException;
import java.io.Serializable;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.HashMap;
import java.util.Map;

import com.googlecode.contraildb.core.IResult;
import com.googlecode.contraildb.core.Identifier;
import com.googlecode.contraildb.core.async.Action;
import com.googlecode.contraildb.core.async.Handler;
import com.googlecode.contraildb.core.async.If;
import com.googlecode.contraildb.core.async.Series;
import com.googlecode.contraildb.core.async.TaskUtils;
import com.googlecode.contraildb.core.storage.provider.IStorageProvider;
import com.googlecode.contraildb.core.utils.ContrailTaskTracker;
import com.googlecode.contraildb.core.utils.ExternalizationManager;
import com.googlecode.contraildb.core.utils.LRUIdentifierIndexedStorage;
import com.googlecode.contraildb.core.utils.tasks.ExternalizationTask;


/**
 * An API for storing Java objects to a raw storage instance.
 * This storage API...
 * 
 * 	...handles the serialization of Java objects to byte streams.
 * 
 * 	...introduces some JPA-like lifecycle management.
 * 		If a stored object implements the Lifecycle interface then 
 * 		the Lifecycle methods will be invoked at appropriate points.
 * 
 * 	...cache objects in order to avoid as much serialization and 
 * 		deserialization as possible (stored objects are cached, but not the lists of children).
 * 
 * 	...can be used by multiple clients in multiple threads, each client should call the
 * 		ObjectStorage.connect method to create its own session.  
 * 		It is safe to use an instance of ObjectStorage from multiple threads. 
 * 		It is also safe to use an instance of ObjectStorage.Session from multiple threads. 
 * 
 * 	...is multithreaded.  This implementation manages the order in which 
 * 		its internal tasks are executed so that operations are performed in a 
 * 		way that allows for as much parallelization as possible in order to 
 * 		maximize performance.  
 * 
 * Like many of Contrail's internal API's all the methods in this class are asynchronous, 
 * they all return an instance of IResult.
 *   
 * @author Ted Stockwell
 */
@SuppressWarnings({"unchecked","rawtypes"})
public class ObjectStorage {
	
	private IStorageProvider _storageProvider;
	private LRUIdentifierIndexedStorage _cache= new LRUIdentifierIndexedStorage();
	private ContrailTaskTracker _tracker= new ContrailTaskTracker();

	/**
	 * @param storageProvider a raw storage provider
	 */
	public ObjectStorage(IStorageProvider storageProvider) {
		this(storageProvider, null);
	}
	public ObjectStorage(IStorageProvider storageProvider, EntityStorage outerStorage) {
		_storageProvider= storageProvider;
	}

	public IResult<Session> connect() throws IOException {
		return new Handler(_storageProvider.connect()) {
			protected IResult onSuccess() throws Exception {
				IStorageProvider.Session session= (IStorageProvider.Session) incoming().getResult();
				return TaskUtils.asResult(
						new Session(_tracker.beginSession(), session));
			};
		}.toResult();
	}
	
	public IResult<Session> connect(final EntityStorage.Session entitySession) {
		return new Handler(_storageProvider.connect()) {
			protected IResult onSuccess() throws Exception {
				IStorageProvider.Session session= (IStorageProvider.Session) incoming().getResult();
				return TaskUtils.asResult(
						new Session(_tracker.beginSession(), session, entitySession));
			};
		}.toResult();
	}
	public IStorageProvider getStorageProvider() {
		return _storageProvider;
	}
	
	
	public class Session {
		
		private IStorageProvider.Session _storageSession;
		private EntityStorage.Session _outerStorage;
		private ContrailTaskTracker.Session _trackerSession;

		public Session(ContrailTaskTracker.Session tracker, IStorageProvider.Session storageProvider) {
			this(tracker, storageProvider, null);
		}
		public Session(ContrailTaskTracker.Session tracker, IStorageProvider.Session storageProvider, EntityStorage.Session outerStorage) {
			_trackerSession= tracker;
			_storageSession= storageProvider;
			_outerStorage= outerStorage; 
		}
		public <T extends Serializable> IResult<Void> store(final Identifier identifier, final T item) 
		{
			final IResult<byte[]>  serializeTask= new ExternalizationTask(item).submit();

			final ILifecycle lifecycle= (item instanceof ILifecycle) ? (ILifecycle)item : null;
			if (lifecycle != null)
				lifecycle.setStorage(_outerStorage);
			
			_cache.store(identifier, item);
			
			return new Series(
					new If(lifecycle != null) {
						protected IResult onSuccess() {
							return lifecycle.onInsert(identifier);
						}
					},
					new Action() {
						protected IResult onSuccess() {
							return _storageSession.store(identifier, serializeTask);
						}
					}).run();
		}

		public IResult<Void> delete(final Identifier path) {
			return new Handler(fetch(path)) {
				protected IResult onSuccess() throws IOException {
					return new Parallel(
					)

					spawn(new Handler(_storageSession.delete(path)) {
						protected IResult onSuccess() throws IOException {
							_cache.delete(path);
							return null;
						}
					});

					Object item= incoming().getResult();
					if (item instanceof ILifecycle) 
						spawn(((ILifecycle)item).onDelete());
					
					return null;
				}
			};
		}

		public <T extends Serializable> IResult<T> fetch(final Identifier path) 
		{
			return new Handler(_cache.fetch(path)) {
				protected IResult onSuccess() throws IOException {
					Object storable= incoming().getResult();
					if (storable != null)
						return TaskUtils.asResult(storable);

					IResult content= _storageSession.fetch(path);
					return readStorable(path, content);
				}
			}.toResult();
		}
		
		private <T extends Serializable> IResult<T> readStorable(final Identifier id, IResult<byte[]> contents)
		throws IOException
		{
			return new Handler(contents) {
				protected IResult onSuccess() throws Exception {
					byte[] bytes= (byte[]) incoming().getResult();
					T s= ExternalizationManager.readExternal(new DataInputStream(new ByteArrayInputStream(bytes)));
					boolean isStorable= s instanceof ILifecycle;
					if (isStorable)
						((ILifecycle)s).setStorage(_outerStorage);
					_cache.store(id, s);
					if (isStorable)
						spawn(((ILifecycle)s).onLoad(id));
					return TaskUtils.asResult(s);
				};
			}.toResult();
		}

		public <T extends Serializable> IResult<Map<Identifier, T>> fetchChildren(final Identifier path)
		{
			return new Handler<Collection<Identifier>,Map<Identifier, T>>(_storageSession.listChildren(path)) {
				protected IResult<Map<Identifier, T>> onSuccess() throws Exception {
					Collection<Identifier> children= incoming().getResult();
					
					HashMap<Identifier, IResult<byte[]>> fetched= new HashMap<Identifier, IResult<byte[]>>();
					for (Identifier childId:children) 
						fetched.put(childId, _storageSession.fetch(childId));
					
					final HashMap results= new HashMap<Identifier, T>();
					for (final Identifier childId:children) {
						spawn(new Handler(readStorable(childId, fetched.get(childId))) {
							protected IResult onSuccess() throws Exception {
								results.put(childId, incoming().getResult());
								return TaskUtils.DONE;
							}
						});
					}
					return TaskUtils.asResult(results);
				};
			}.toResult();
		}

		public IResult<Collection<Identifier>> listChildren(final Identifier path)
		{
			return _storageSession.listChildren(path);
		}
		
		public IResult<Void> flush() {
			return new Handler(_trackerSession.complete()) {
				protected IResult onSuccess() throws Exception {
					spawn(_storageSession.flush());
					return TaskUtils.DONE;
				}
			}.toResult();
		}
		
		public IResult<Void> close() {
			return new Handler(flush()) {
				protected IResult onSuccess() throws Exception {
					spawn(new Handler(_trackerSession.close()) {
						protected IResult onSuccess() throws Exception {
							spawn(new Handler(_storageSession.close()) {
								protected IResult onSuccess() throws Exception {
									_trackerSession= null;
									_storageSession= null;
									_outerStorage= null;
									return TaskUtils.DONE;
								};
							});
							return TaskUtils.DONE;
						};
					});
					return TaskUtils.DONE;
				};
			}.toResult();
		}
		
		public <T extends Serializable> IResult<Boolean> create(final Identifier identifier, final T item, final long waitMillis)
		{
			final IResult<byte[]> serializeTask= new ExternalizationTask(item).submit();
			return new Handler(_storageSession.create(identifier, serializeTask, waitMillis)) {
				protected IResult onSuccess() throws Exception {
					boolean isStorable= item instanceof ILifecycle;
					if (isStorable)
						((ILifecycle)item).setStorage(_outerStorage);
					_cache.store(identifier, item);
					if (isStorable) 
						spawn(((ILifecycle)item).onInsert(identifier));
					return TaskUtils.DONE;
				};
			}.toResult();
		}
		
		
		public IResult<Void> delete(Identifier... paths) {
			ArrayList<IResult> tasks= new ArrayList<IResult>();
			for (Identifier identifier:paths)
					tasks.add(delete(identifier));
			return TaskUtils.combineResults(tasks);
		}

		public IResult<Void> deleteAllChildren(Identifier... paths) {
			return deleteAllChildren(Arrays.asList(paths));
		}

		public IResult<Void> deleteAllChildren(Iterable<Identifier> paths) {
			ArrayList<IResult> tasks= new ArrayList<IResult>();
			for (Identifier identifier:paths)
				deleteAllChildren(identifier);
			return TaskUtils.combineResults(tasks);
		}

		public IResult<Void> deleteAllChildren(Identifier path) {
			final IResult<Collection<Identifier>> children= listChildren(path);
			return new Handler(children) {
				protected IResult onSuccess() throws Exception {
					ArrayList<IResult> tasks= new ArrayList<IResult>();
					for (Identifier identifier: children.get()) {
						tasks.add(delete(identifier));
					}
					return TaskUtils.combineResults(tasks);
				};
			}.toResult();
		}


	}
}

File: contrail/src/contrail/storage/ObjectStorage.go
package storage

import (
	"bytes"
	"time"
	"encoding/gob"
	"contrail/id"
	"contrail/tasks"
)

/**
 * An API for storing objects to a raw storage instance.
 * This storage API...
 * 
 * 	...handles the serialization of objects to byte streams.
 * 
 * 	...introduces some JPA-like lifecycle management.
 * 		If a stored object implements the Lifecycle interface then 
 * 		the Lifecycle methods will be invoked at appropriate points.
 * 
 * 	...caches objects in order to avoid as much serialization and 
 * 		deserialization as possible.
 * 
 * 	...can be used by multiple clients in multiple threads.
 *		Each client should call the ObjectStorage.Connect method to create its 
 *		own session.  
 * 		It is safe to use an instance of ObjectStorage from multiple threads. 
 * 		It is also safe to use an instance of ObjectStorage.Session from 
 * 		multiple threads. 
 * 
 * 	...is parallel.  Each session manages the order in which its internal tasks 
 *		are executed so that operations are performed in a way that allows for 
 *		as much parallelization as possible in order to maximize performance.  
 *   
 * @author Ted Stockwell
 */
type ObjectStorage struct {
	storageProvider StorageProvider
}

type ObjectStorageSession struct {
	owner *ObjectStorage
	storageSession StorageSession
	cache id.TreeStorage
	taskMaster *tasks.TaskMaster
}

func CreateObjectStorage (storageProvider StorageProvider) *ObjectStorage {
	return &ObjectStorage { storageProvider: storageProvider }
}

func (self *ObjectStorage) Connect() *ObjectStorageSession {
	return &ObjectStorageSession {
		owner: self,
		storageSession: self.storageProvider.Connect(),
		cache:  id.CreateLRUTreeStorage(cacheSize),
		taskMaster: tasks.CreateTaskMaster(),
	}
}

func (self *ObjectStorageSession) Store(identifier *id.Identifier, item interface{}) {

	serializeTask:= tasks.GoResult(func() interface{} { return SerializeObject(item) })
	
	lifecycle:= item.(Lifecycle)
	if lifecycle != nil {
		lifecycle.SetStorage(self)
	}

	self.taskMaster.Submit(tasks.WRITE, identifier, func() interface{} {
		self.cache.Store(identifier, item);
		if (lifecycle != nil) {
			lifecycle.OnInsert(identifier)
		}
		self.storageSession.Store(identifier, serializeTask.Get().([]byte))
		return nil
	})
}

func (self *ObjectStorageSession) Delete(path *id.Identifier) {
	self.taskMaster.Submit(tasks.DELETE, path, func() interface{} {
		item:= self.Fetch(path)
		self.cache.Delete(path)
		self.storageSession.Delete(path)
		lifecycle:= item.(Lifecycle)
		if lifecycle != nil {
			lifecycle.OnDelete()
		}
		return nil
	})
}

func (self *ObjectStorageSession) Fetch(path *id.Identifier) interface{} {
	return self.taskMaster.Submit(tasks.READ, path, func() interface{} {
		storable:= self.cache.Fetch(path)
		if storable == nil {
			bytes:= self.storageSession.Fetch(path)
			if bytes != nil { 
				storable= self.readStorable(path, bytes)
			}
		}
		return storable
	}).Get()
}

func (self *ObjectStorageSession) readStorable(path *id.Identifier, contents []byte) interface{} {
	if contents == nil { return nil }

	storable:= DeserializeObject(contents)
	if storable == nil {  return nil }
	
	lifecycle:= storable.(Lifecycle)
	if lifecycle != nil { lifecycle.SetStorage(self) }
	
	self.cache.Store(path, storable)
	
	if lifecycle != nil { lifecycle.OnLoad(path) }
	return storable
}

func (self *ObjectStorageSession) FetchChildren(path *id.Identifier) id.TreeStorage {
	return self.taskMaster.Submit(tasks.LIST, path, func() interface{} {
		results:= id.CreateTreeStorage()
		children:= self.storageSession.ListChildren(path)
		todo:= make([]*tasks.Future, len(children))
		for i,childId:= range children {
			todo[i]= tasks.Go(func() {
				bytes:= self.storageSession.Fetch(childId)
				object:= self.readStorable(childId, bytes)
				results.Store(childId, object)
			})
		}
		tasks.WaitAll(todo)
		return results
	}).Get().(id.TreeStorage)
}


/*
 * Implementation Note:
 * Caching is not problematic, even though there can be multiple processes writing 
 * to storage, since the higer level Contrail APIs never changes a stored object, 
 * it only writes new versions.  
 * However, while caching stored objects is not a problem, caching lists of children is,
 * since obviously other processes are adding children. 
 */
func (self *ObjectStorageSession) ListChildren(path *id.Identifier) []*id.Identifier {
	return self.taskMaster.Submit(tasks.LIST, path, func() interface{} {
		return self.storageSession.ListChildren(path)
	}).Get().([]*id.Identifier)
}

func (self *ObjectStorageSession) Flush() {
	self.taskMaster.Join()
	self.storageSession.Flush()
}

func (self *ObjectStorageSession) Close() {
	self.Flush()
	self.taskMaster.Close()
	self.storageSession.Close()
	
	self.storageSession= nil
	self.taskMaster= nil
	self.cache= nil
}

func (self *ObjectStorageSession) Create(path *id.Identifier, item interface{}, wait time.Duration) bool {
	serializeTask:= tasks.GoResult(func() interface{} { return SerializeObject(item) })
	return self.taskMaster.Submit(tasks.LIST, path, func() interface{} {
		created:= false;
		if self.storageSession.Create(path, serializeTask.Get().([]byte), wait) {
			lifecycle:= item.(Lifecycle)
			if lifecycle != nil {
				lifecycle.SetStorage(self)
			}
			self.cache.Store(path, item)
			if lifecycle != nil {
				lifecycle.OnInsert(path)
			}
			created= true;
		}
		return created
	}).Get().(bool)
}


func (self *ObjectStorageSession) DeleteAll(paths []*id.Identifier) {
	todo:= make([]*tasks.Future, len(paths))
	for i,path:= range paths {
		todo[i]= tasks.Go(func() { self.Delete(path) })
	}
	tasks.WaitAll(todo)
}

func (self *ObjectStorageSession) DeleteAllChildren(path *id.Identifier) {
	children:= self.ListChildren(path);
	todo:= make([]*tasks.Future, len(children))
	for i,path:= range children {
		todo[i]= tasks.Go(func(){ self.Delete(path) })
	}
	tasks.WaitAll(todo)
}

func SerializeObject(item interface{}) []byte {
    buffer:= new(bytes.Buffer)
    enc := gob.NewEncoder(buffer) 
    err := enc.Encode(item)
    if err != nil {
        panic(err)
    }
    return buffer.Bytes()
}


func DeserializeObject(data []byte) interface{} {
    buffer:= bytes.NewBuffer(data)
    dec := gob.NewDecoder(buffer) 
    var item interface{}
    err:= dec.Decode(item)
    if err != nil {
        panic(err)
    }
    return item	
}